logstreams/src/main/java/io/zeebe/logstreams/processor/ReProcessingStateMachine.java

```diff
@@ -118,53 +113,107 @@ private ReProcessingStateMachine(
     this.streamProcessorName = context.getName();
     this.eventFilter = context.getEventFilter();
     this.logStreamReader = context.getLogStreamReader();
+    this.isReadOnlyProcessor = context.isReadOnlyProcessor();
+    this.producerId = context.getId();
 
     this.streamProcessor = streamProcessor;
     this.zeebeDb = zeebeDb;
-    this.updateStateRetryStrategy = new RecoverableRetryStrategy(actor);
+    this.updateStateRetryStrategy = new EndlessRetryStrategy(actor);
+    this.processRetryStrategy = new EndlessRetryStrategy(actor);
     this.abortCondition = abortCondition;
   }
 
   // current iteration
   private long lastSourceEventPosition;
   private ActorFuture<Void> recoveryFuture;
-
   private LoggedEvent currentEvent;
   private EventProcessor eventProcessor;
   private ZeebeDbTransaction zeebeDbTransaction;
 
-  ActorFuture<Void> startRecover(long lastSourceEventPosition) {
-    this.lastSourceEventPosition = lastSourceEventPosition;
+  ActorFuture<Void> startRecover(final long snapshotPosition) {
     recoveryFuture = new CompletableActorFuture<>();
-    reprocessNextEvent();
+
+    final long startPosition = logStreamReader.getPosition();
+
+    LOG.info("Start scanning the log for error events.");
+    lastSourceEventPosition = scanLog(snapshotPosition);
+    LOG.info(
+        "Processor {} starts reprocessing, until last source event position {}",
+        streamProcessorName,
+        lastSourceEventPosition);
+
+    if (lastSourceEventPosition > snapshotPosition) {
+      logStreamReader.seek(startPosition);
+      reprocessNextEvent();
+    } else {
+      recoveryFuture.complete(null);
+    }
     return recoveryFuture;
   }
 
-  private void reprocessNextEvent() {
-    try {
-      if (!logStreamReader.hasNext()) {
-        throw new IllegalStateException(
-            String.format(
-                ERROR_MESSAGE_REPROCESSING_NO_NEXT_EVENT,
-                lastSourceEventPosition,
-                streamProcessorName));
+  private long scanLog(final long snapshotPosition) {
+    long lastSourceEventPosition = -1L;
+
+    if (!isReadOnlyProcessor && logStreamReader.hasNext()) {
+      lastSourceEventPosition = snapshotPosition;
+      while (logStreamReader.hasNext()) {
+        final LoggedEvent newEvent = logStreamReader.next();
+
+        final long errorPosition = streamProcessor.getFailedPosition(newEvent);
+        if (errorPosition >= 0) {
+          LOG.info(
```

Comment:
We should change the log level to `TRACE` because the statement is mainly to understand the behavior while developing.

Reply:
I would vote against putting it on trace, this is also an indication to users that something is going wrong.

Reply:
No, it's a log statement that something went wrong on processing. On re-processing, we just see that an error happens on processing :thinking:

Reply:
So ? Meeting in the middle ? (debug) :sweat_smile: